### ISSUE TITLE
Parameter to block is shadowed by private varaible

### ISSUE THREAD

**Issue by swiperthefox** [jmoenig/Snap#2572]
Let's say I have a sprite `Worker`, and it has a private variable `age`. I want to make instances of `Worker` with different ages.

A `create worker` block is like the following:

```
 create worker (age):
   set new_worker = a clone of Worker
   tell new_worker to {set age = age}
   report new_worker
```

But this new block will not work, the `age` of the new clone is always the same as that of the original one (but it is a copied one since changing the `age` of the original does not affect the clone's `age`).

If we change the parameter name `age` to `worker_age` as shown below, then it will work as expected. It seems that the `age` as a parameter is shadowed by the private property `age`. This surprised me as we commonly see `this.age = age` or `self.age = age`.

```
 create worker (worker_age):
   set new_worker = a clone of Worker
   tell new_worker to {set age = worker_age}
   report new_worker
```

Here is a demo of the problem. [initialize clone](https://snap.berkeley.edu/snap/snap.html#present:Username=swiperthefox&ProjectName=initializeclone)

**Comment by brianharvey** [comment]
What? (a) What's it reserved for? (b) If it really is, that's a misfeature that should be repaired; there shouldn't be reserved words.  And (c) if I edit the project to change every instance of "name" to "bloop," it still fails in the same way.

Apparently the sprite variable shadows the script variable, instead of the other way around.  This is a bug imho.

**Comment by jguille2** [comment]
Hi Brian (@brianharvey )

I think it is not a bug. It's the opposite: the script variable is shadowing the sprite one.

Example. If I have different sprites with a "sprites variable" named "param".
 - If, from the sprite1, I did

![set1](https://user-images.githubusercontent.com/2926315/79184398-4fb61480-7e14-11ea-9978-8d895847e2a5.png)

Then, I will change Sprite(2)'s "param".

- But If, again from Sprite1, I did:

![set2](https://user-images.githubusercontent.com/2926315/79184469-8429d080-7e14-11ea-941e-79f321d6adb5.png)

Then, I will change nothing. Because in this scope, calling "param" is calling the script variable.

And I think this is fine.
Joan

**Comment by jmoenig** [comment]
yes, the parameter eclipses the sprite-local var.

@brianharvey Did you look at @swiperthefox's script at all before reopening this issue? 

There are 2 obvious problems with the example script, and one that's slightly less obvious. 

What does anybody expect this statement to do:

![initializeclone script pic](https://user-images.githubusercontent.com/2596285/79192131-72522880-7e28-11ea-848e-d3871d458114.png)

If the answer is anything but "nothing" we can stop this discussion right here.

So we can instead speculate what this script *wants* to accomplish and whether that's possible at all. 

The second obvious issue is that parameters are mutable in Snap! and their scope eclipses sprite-local and global vars as it should be. Even if we fix this there is still a possible collision in the `set` block because there also is an attribute called "name". That's the not-so-obvious third issue that might be encountered once the two previous obvious ones have been taken care of. I was assuming everybody to already be at the third step, because the two other ones are ... obvious.

TL;DR: Don't name everything the same and then complain it's a bug in Snap!

**Comment by brianharvey** [comment]
OOOOHHHH!  I was focused on the second input to the SET, so I was thinking it'd have to be
![untitled script pic](https://user-images.githubusercontent.com/3885074/79198218-ffc15880-7de7-11ea-93f8-159bb3f57e08.png)
but (of course, I now see) the problem was with the _first_ input to SET!  Hmm, I just had the thought that maybe I could make it work with LAUNCH, but no, its scope still includes the script variable (the parameter).  I can't see any way to make it work, other than using a different parameter -- and you're right, that'd be good practice anyway, even if it _could_ work.

But you can't say SET MY>NAME TO ... and so users have no way of knowing not to call something "name," which is (as in this project) a very natural variable name to want to use.  In general I think we shouldn't have reserved variable names -- and now I remember why I wanted a SET MY block in Sensing accompanying the MY reporter block, instead of that submenu in SET.  Besides NAME, other things in the MY reporter menu that shouldn't be off limits to user variable names are LEFT and RIGHT.  (TOP and BOTTOM don't bother me quite as much because I'd be more inclined to use UP and DOWN as variable names, but in principle there should be no reserved names at all, not even "rotation x."

Anyway, sorry about not being up to speed; I'm getting rusty!  I'll re-close this, but I'm going to argue for changing to SET MY in Sensing.

**Comment by jmoenig** [comment]
naming a sprite-var *"name"* is not off-limits. You can have a local variable called "name" and set it using the `set` block, it's just that *looking* at the `set` block won't give you a visual clue as to whether it's accessing the built-in attribute or the user-defined variable (which is something we should probably change, you're right), but it totally works (for non-temporary clones).

**Comment by brianharvey** [comment]
Ah.  Cool.  But I now don't feel guilty about misunderstanding "Name is a reserved property" above with no context!  :-)
